compiler/rustc_middle/src/ty/trait_def.rs

```diff
@@ -147,8 +143,24 @@ impl<'tcx> TyCtxt<'tcx> {
         self_ty: Ty<'tcx>,
     ) -> impl Iterator<Item = DefId> + 'tcx {
         let impls = self.trait_impls_of(def_id);
-        if let Some(simp) = fast_reject::simplify_type(self, self_ty, TreatParams::AsInfer) {
-            if let Some(impls) = impls.non_blanket_impls.get(&simp) {
+
+        // Non-blanket impls are grouped into two different maps:
+        // 1) impls for references to simplifiable types
+        if let TyKind::Ref(_, ref_ty, _) = self_ty.kind() {
+            if let Some(simplified_ref_ty) =
+                fast_reject::simplify_type(self, *ref_ty, TreatParams::AsInfer)
+            {
+                if let Some(impls) = impls.impls_for_ref_x.get(&simplified_ref_ty) {
+                    return impls.iter().copied();
+                }
+            }
```

Comment:
missing the `impls_for_ref_x` impls if you have `&T`. Doesn't matter with how this method is used but we should still be careful for it to be correct.